pennylane/transforms/compile.py

```diff
@@ -0,0 +1,166 @@
+
+    We can compile it down to a smaller set of gates using the ``qml.compile``
+    transform.
+
+    >>> compiled_qfunc = qml.compile()(qfunc)
+    >>> compiled_qnode = qml.QNode(compiled_qfunc, dev)
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──H───RX(0.6)───────────────────┤ ⟨Z⟩
+     1: ──H──╭X─────────────────╭CY─────┤
+     2: ──H──╰C────────RX(0.3)──╰CY──Y──┤
+
+    You can change up the set of transforms by passing a custom ``pipeline`` to
+    ``qml.compile``. The pipeline is a list of transform functions. Furthermore,
+    you can specify a number of passes (repetitions of the pipeline), and a list
+    of gates into which the compiler will first attempt to decompose the
+    existing operations prior to applying any optimization transforms.
+
+    .. code-block:: python3
+
+        compiled_qfunc = qml.compile(
+            pipeline=[
+                qml.transforms.commute_controlled(direction="left"),
+                qml.transforms.merge_rotations(atol=1e-6),
+                qml.transforms.cancel_inverses
+            ],
+            basis_set=["CNOT", "RX", "RY", "RZ"],
+            num_passes=2
+        )(qfunc)
+
+        compiled_qnode = qml.QNode(compiled_qfunc, dev)
+
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──RZ(1.57)──RX(1.57)──RZ(1.57)───RX(0.6)───────────────────────────────────────────────────────────────────────┤ ⟨Z⟩
+     1: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╭X────────RZ(1.57)──────────────────────────────────────────╭C─────────────╭C──┤
+     2: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╰C────────RX(0.3)───RZ(1.57)──RY(3.14)──RZ(1.57)──RY(1.57)──╰X──RY(-1.57)──╰X──┤
+
+    """
+    # Ensure that everything in the pipeline is a valid qfunc or tape transform
+    if pipeline is None:
+        pipeline = default_pipeline
+    else:
+        for p in pipeline:
+            p_func = p.func if isinstance(p, partial) else p
+            if not isinstance(p_func, single_tape_transform) and not hasattr(p_func, "tape_fn"):
+                raise ValueError("Invalid transform function {p} passed to compile.")
+
+    if basis_set is not None:
+        for gate_name in basis_set:
+            if gate_name not in all_ops or not isinstance(gate_name, str):
+                raise ValueError("Invalid basis gate {gate_name} passed to compile.")
+
+    if num_passes < 1 or not isinstance(num_passes, int):
+        raise ValueError("Number of passes must be an integer with value at least 1.")
+
+    # Expand the tape; this is done to unroll any templates that may be present,
+    # as well as to decompose over a specified basis set
+    # First, though, we have to stop whatever tape may be recording so that we
+    # don't queue anything as a result of the expansion or transform pipeline
+    current_tape = get_active_tape()
+
+    with current_tape.stop_recording():
+        if basis_set is not None:
+            expanded_tape = tape.expand(depth=5, stop_at=lambda obj: obj.name in basis_set)
```

Comment:
how come 5?

Reply:
Perhaps this should be added as a keyword argument? Just in case someone wants to modify this in the future?